components/nexus-repository/src/main/java/org/sonatype/nexus/repository/storage/OrientAsyncHelper.java

```
@@ -0,0 +1,137 @@
+import com.orientechnologies.orient.core.db.document.ODatabaseDocumentTx;
+import com.orientechnologies.orient.core.record.impl.ODocument;
+import com.orientechnologies.orient.core.sql.query.OSQLNonBlockingQuery;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+/**
+ * Helper class to wrap Orient SQL SELECT queries into {@link Iterable}s.
+ *
+ * @since 3.0
+ */
+class OrientAsyncHelper
+    extends ComponentSupport
+{
+  private static final int BUFFER_SIZE = 128;
+
+  private static final long TIMEOUT_SECONDS = 60L;
+
+  /**
+   * Executes with passed in {@link ODatabaseDocumentTx} connection the passed in {@link String} SELECT SQL command
+   * and returns a true asynchronous backed {@link Iterable} of {@link ODocument} instance, that can be later
+   * transformed into whatever caller needs. This means that this wrapper supports ANY SQL commands, even those
+   * performing aggregation on results, as long as iterator is not accumulated into a real array or list. Still, given
+   * the nature of non-blocking query, it's execution happens on a separate thread, hence, it will not see uncommited
+   * changes from the caller thread, if any.
+   */
+  public static Iterable<ODocument> asyncIterable(final ODatabaseDocumentTx db,
+                                                  final String selectQuery,
+                                                  @Nullable final Map<String, Object> parameters)
+  {
+    checkNotNull(db);
+    checkNotNull(selectQuery);
+    final ArrayBlockingQueue<ODocument> queue = new ArrayBlockingQueue<>(BUFFER_SIZE);
+
+    db.command(new OSQLNonBlockingQuery<ODocument>(selectQuery, new OCommandResultListener()
+    {
+      @Override
+      public boolean result(final Object iRecord) {
+        final ODocument doc = (ODocument) iRecord;
+        try {
+          queue.offer(doc, TIMEOUT_SECONDS, TimeUnit.SECONDS);
+        }
+        catch (InterruptedException e) {
+          throw Throwables.propagate(e);
```

I think it would be safer to log the exception and return `false` to explicitly cancel the query, just in case the calling code swallows the exception or gets into an odd state.